proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1155 @@
+      visibility override
+    * it does not move us towards a single syntax for types and terms, which would
+      be an advantage when we have dependent types
+    * The dual purpose of ``@`` as both a visibility override and a namespace
+      specifier would lead to unwanted interference between ``forall x.`` and
+      ``forall x ->``. For example, given ``f :: forall k. forall (a::k) ->
+      blah``, it wouldn't be possible to specify ``a=Int`` as ``f @Int``;
+      one would have to write ``f @_ @Int`` or change the type of ``f`` to ``f
+      :: forall {k}. forall (a::k) -> blah``.
+
+    Richard Eisenberg characterizes this alternative as follows:
+
+      It moves us away from uniformity. Let's even pretend for a moment that
+      I'm not trying to actually merge the term-level and type-level.
+
+      Right now, we can say this::
+
+        type VDQ :: forall k1. forall k2 -> k1 -> k2 -> Type
+        data VDQ k2 a b
+
+        type VDQIntTrue = VDQ @Type Bool Int True
+        type VDQCharFalse = VDQ Bool Char False
+
+      If we were to require the ``@`` in terms, the term-level equivalent would be::
+
+        vid :: forall a. forall b -> a -> b -> ()
+        vid _ _ _ = ()
+
+        ex1 = vid @Int @Bool 3 True
+        ex2 = vid @_ @Bool 'x' False
+
+      These look different! Why different syntaxes for the same idea?
+
+      Worse, imagine a data constructor::
+
+        data Silly a b where
+          Mk :: forall a. forall b -> a -> b -> Silly a b
+
+      Now we have this oddity::
+
+        type Different1 = Mk @Nat Bool 3 True
+        type Different2 = Mk Bool "hi" False
+        different3 = Mk @Int @Bool 3 True
+        different4 = Mk @_ @Bool "hi" False
+
+      Here, the right-hand sides should be *the same*, but they have to be
+      different.
+
+      Today, we have non-uniformity by omission: we have no visible ``forall`` in
+      types of terms. But with your proposed ``@`` on required dependent
+      arguments, we would have active non-uniformity, which seems worse as it
+      paints us into a corner that's difficult to escape from. At least
+      non-uniformity by omission can, in theory, be fixed uniformly, later.
+
+6. A previous iteration of this proposal dictated to switch to a type-level
+   name resolution context when processing types-in-terms; we could also parse
+   the right-hand side of ``forall a. t`` as a type; and we could map the
+   ``forall`` in terms bind variables in the type namespace.
+
+   The parsing and name resolution rules of these alternatives were deemed to
+   subtle, so we opted for a design where types-in-terms are parsed and renamed
+   as ordinary terms.
+
+7. We could error on ambiguous variable occurrences earlier in the pipeline, in
+   the renamer, but then enabling ``RequiredTypeArguments`` would result in
+   rejecting currently valid code::
+
+    id :: forall a. a -> a
+    id a = (a :: a)
+
+   Instead, we opted to raise the ambiguity error during T2T.
```

Comment:
There is an alternative missing: don't include any type syntax in terms, and always force the `type` herald where a visible type argument would be.

The reason why I'm raising this is: I don't know how much of the proposal we could remove if we removed the ability to not use the `type` herald. I realise that you are not particularly interested in a proposal which forces the `type` herald in this case, but, as a committee member, I do find it hard to evaluate the proposal without this knowledge (even if it's not fully precise).

Reply:
I'd like to second this request. There seem to be two related, but separate proposals here:

1. Allow visible `forall` in types of terms, with some sort of syntactic herald (whether `@` or `type`, etc) to allow the injection of type syntax into term syntax.

2. A partial translation of term syntax to type syntax to allow the omission of the `type` herald, which I agree makes the extension more ergonomic.

Like @yav, I found the proposal to be substantially more complex than I expected, and I believe it's largely a result of (2). It would be helpful if the proposal separated the two pieces so we could evaluate them independently. In particular, I think (1) would be a valuable addition regardless of the discussion around Dependent Haskell.